### PR TITLE
service controller trigger update when service ipfamilies changes

### LINKED ISSUE #116495: [dual-stack]support service ipFamilies update event in service controller
### What would you like to be added?

Currently, when loadbalancer service's ipFammilies updated，service controller will ignore this event. That means cloud provider can't update ipv4/ipv6 backend.
So, I'm thinking service controller need add watching 'service's ipFammilies updated' event, It'd be important for cloud provider update dual-stack backend.

code at :https://github.com/kubernetes/cloud-provider/blob/fa271f103b481b342ed5eefd9cde68c30bf3b114/controllers/service/controller.go#L526

### Why is this needed?

Our load balancer supports dual-stack. When the services in the cluster are updated from IPv4 to IPv6(same ipv6->ipv4、ipv4->ipv4/ipv6), it is expected that the cloud provider can perceive this event and update the IPv4/IPv6 address to the backend of the load balancer.

/sig cloud-provider
/area cloudprovider

### PR DESCRIPTION
**Pull request by Dingshujie**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
when loadbancer service ipFamilies change from dualstack to single stack or from single stack to dualstack，service controller need to trigger update in cloud provider

#### Which issue(s) this PR fixes:

Fixes #116495

#### Does this PR introduce a user-facing change?

```release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by panslava** [comment]
/lgtm

seems fine, only small thing, that it strongly relies on the fact that first ipFamily can not change, which is true, and probably will continue being true

**Comment by Dingshujie** [comment]
> /lgtm
> 
> seems fine, only small thing, that it strongly relies on the fact that first ipFamily can not change, which is true, and probably will continue being true

[service update validate L5176](https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/validation/validation.go#L5176) defines User can upgrade (add another clusterIP or ipFamily) or can downgrade (remove secondary clusterIP or ipFamily), but *CAN NOT* change primary/secondary clusterIP || ipFamily *UNLESS* they are changing from/to/ON ExternalName

File: staging/src/k8s.io/cloud-provider/controllers/service/controller.go [hunk 1]
```diff
@@ -582,6 +582,12 @@ func (c *Controller) needsUpdate(oldService *v1.Service, newService *v1.Service)
 		return true
 	}
 
+	if len(oldService.Spec.IPFamilies) != len(newService.Spec.IPFamilies) {
```
Comment (andrewsykim) [inline staging/src/k8s.io/cloud-provider/controllers/service/controller.go:585]: Does the order matter in this case? Should we check order in addition to length of ip families?

Comment (aojea) [reply staging/src/k8s.io/cloud-provider/controllers/service/controller.go:585]: see comment https://github.com/kubernetes/kubernetes/pull/116520#issuecomment-1467232074

Comment (andrewsykim) [reply staging/src/k8s.io/cloud-provider/controllers/service/controller.go:585]: can we add a comment in the code with this context?

Comment (andrewsykim) [reply staging/src/k8s.io/cloud-provider/controllers/service/controller.go:585]: can we add a comment in the code with this context?

Comment (Dingshujie) [reply staging/src/k8s.io/cloud-provider/controllers/service/controller.go:588]: done, add comment with this context. PTAL

**Review by andrewsykim — COMMENTED** [review]
/approve
/lgtm

**Comment by Dingshujie** [comment]
@andrewsykim @aojea PLAT

**Comment by dims** [comment]
re-apply LGTM (after the gofmt update)

/lgtm

**Pull request merged**